### PR TITLE
[CDAP-19949] Add retries for sql tx connection failures

### PR DESCRIPTION
**Pull request by itsankit-google**
[CDAP-19949](https://cdap.atlassian.net/browse/CDAP-19949)

### CODE REVIEW

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java [hunk 1]
```diff
@@ -72,25 +82,40 @@ public void run(TxRunnable runnable) throws TransactionException {
...
-              Thread.currentThread().interrupt();
-              // Fail with the original exception
-              throw e;
-            }
-          }
+          applyDelay(retries, delayMillisTransactionFailure, e);
         } else {
           throw e;
         }
+      } catch (TransactionException e) {
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java]: Having two catch blocks for SqlTransactionException and Transactionexception looks a bit strange. Looking at SqlTransactionRunner#run when the cause is SqlException you should be getting SqlTransactionException every time the cause is SQLException, should not you?

**Comment by tivv** [comment]
Also please create a test using embedded postgres (see io.cdap.cdap.spi.data.sql.SqlStructuredTableTest as example)

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java (hunk 1 above)
Comment (itsankit-google) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java]: I thought about it earlier, so there are two places where connection failure can occur - 
1. https://github.com/cdapio/cdap/blob/fab9cfd6b79cf3f23f74fa272771e5434b44d8c4/cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java#L62
2. https://github.com/cdapio/cdap/blob/fab9cfd6b79cf3f23f74fa272771e5434b44d8c4/cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java#L66-L72

At the first place, we don't throw a sql transaction exception so I had to handle it like this.
wdyt??

Comment (itsankit-google) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java]: Discussed offline to throw sql transaction exception at both places.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java [hunk 2]
```diff
@@ -55,6 +60,11 @@ public RetryingSqlTransactionRunner(StructuredTableAdmin tableAdmin, DataSource
                                cConf.getBoolean(Constants.Metrics.STRUCTURED_TABLE_TIME_METRICS_ENABLED),
                                scanFetchSize);
     this.metricsCollectionService = metricsCollectionService;
+    this.maxRetries = cConf.getInt(Constants.Dataset.DATA_STORAGE_SQL_TRANSACTION_RUNNER_MAX_RETRIES, 20);
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java]: Don't add defaults here. Defaults should come from cdap-default.xml

Comment (itsankit-google) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/RetryingSqlTransactionRunner.java]: ack

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java [hunk 3]
```diff
@@ -59,7 +59,7 @@ public void run(TxRunnable runnable) throws TransactionException {
     try {
       connection = dataSource.getConnection();
     } catch (SQLException e) {
-      throw new TransactionException("Unable to get connection to the sql database", e);
+      throw new SqlTransactionException(e, "Unable to get connection to the sql database.");
```
Comment (tivv) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java]: nit: Follow the regular exception constructor code style: message first, exception second

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableRetryTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    SqlTransactionRunner mockedSqlTransactionRunner =
+      Mockito.spy(((RetryingSqlTransactionRunner) transactionRunner).getSqlTransactionRunner());
+
+    Mockito.doAnswer(i -> {
+      if (failedCalls.isEmpty()) {
+        // Fail first transaction with sql transaction failure.
+        failedCalls.add(1);
+        throw new SqlTransactionException(new SQLException(SQL_TRANSACTION_EXCEPTION, TRANSACTION_CONFLICT_SQL_STATE),
+                                          SQL_TRANSACTION_EXCEPTION_MESSAGE);
+      } else if (failedCalls.contains(1)) {
```
Comment (tivv) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableRetryTest.java]: I believe you don't need a counter.  You can just do 
```
Mockito.doAnswer(firstAnswer).doAnswer(secondAnswer).doAnswer(thirstAnswer).when...
```
see https://www.javadoc.io/doc/org.mockito/mockito-core/2.3.4/org/mockito/stubbing/Stubber.html.

Not sure if it would work for repeated scenarios (1,2,3,1,2,3,...). If so, I don't think you need the whole HashSet. It will alsways have 0/1 items. You can just use AtomitInteger or similar.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java [hunk 5]
```diff
@@ -67,6 +70,15 @@ protected void configure() {
     Assert.assertEquals(RetryingSqlTransactionRunner.class, transactionRunner.getClass());
   }
 
+  /**
+   * populate retry settings in cConf used by {@link SqlStructuredTableRetryTest}
+   */
+  private static void populateCConf(CConfiguration cConf) {
+    cConf.setInt(Constants.Dataset.DATA_STORAGE_SQL_TRANSACTION_RUNNER_MAX_RETRIES, 2);
+    cConf.setLong(Constants.Dataset.DATA_STORAGE_SQL_TRANSACTION_RUNNER_TRANSACTION_FAILURE_DELAY_MILLIS, 1);
```
Comment (tivv) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java]: I believe 0 would work for the test.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java [hunk 6]
```diff
@@ -80,7 +92,7 @@ protected StructuredTableAdmin getStructuredTableAdmin() {
   }
 
   @Override
-  protected TransactionRunner getTransactionRunner() {
+  protected TransactionRunner getTransactionRunner() throws TransactionException {
```
Comment (tivv) [inline cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java]: Why this change?

**Review by tivv — APPROVED** [review]
LGTM with minor comments

File: cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java (hunk 3 above)
Comment (itsankit-google) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/spi/data/sql/SqlTransactionRunner.java]: Ack

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableRetryTest.java (hunk 4 above)
Comment (itsankit-google) [reply cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableRetryTest.java]: Thanks for the pointer, updated. It worked as expected.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java (hunk 6 above)
Comment (itsankit-google) [reply cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java]: Added a try-catch to handle it coming from `sqlTransactionRunner.run()`.

File: cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java (hunk 5 above)
Comment (itsankit-google) [reply cdap-data-fabric/src/test/java/io/cdap/cdap/spi/data/sql/SqlStructuredTableTest.java]: Ack
